### ISSUE TITLE
Pytest fixtures are inherited incorrectly

### ISSUE THREAD

**Issue by clockdev** [pytest-dev/pytest#5573]
Hello! My trouble at the StackOverflow: https://stackoverflow.com/questions/56884929/pytest-fixtures-inherited-everywhere-implicitly
Full code: [gist.github](https://gist.github.com/clockdev/7324f8768ebeabe4abc2781488d83fa0)
Conftest: [gist.github](https://gist.github.com/clockdev/156026bd34147ed32fbcef2711d39f11)

What I want to get: check API with existing permissions and nonexistent.
What I get:
```
(tests from top to bottom)
test_with_permission_fixture
> response code 200
test_without_fixture
> response code 200

(tests from top to bottom)
test_without_fixture
> response code 403
test_with_permission_fixture
> response code 403
```

```
> pytest --version
This is pytest version 4.6.4, imported from 
.../venv/lib/python2.7/site-packages/pytest.pyc
setuptools registered plugins:
  celery-4.2.1 
  pytest-cov-2.7.1 
  pytest-django-3.4.8
  pytest-forked-1.0.2
  pytest-mock-1.10.4
  pytest-xdist-1.29.0
  pytest-xdist-1.29.0
```
```
> pip list

Package                       Version     Location                                                            
----------------------------- ----------- ------------------------------------------------------
acme                          0.4.1       
alabaster                     0.7.12      
amqp                          2.2.2       
apipkg                        1.5         
asgi-ipc                      1.3.0       
asgi-rabbitmq                 0.5.5       
asgi-redis                    1.4.3       
asgiref                       1.1.2       
asn1crypto                    0.22.0      
atomicwrites                  1.3.0       
attrs                         18.2.0      
autobahn                      19.2.1      
Automat                       0.7.0       
autopep8                      1.2.4       
Babel                         2.6.0       
backports-abc                 0.5         
backports.functools-lru-cache 1.5         
backports.ssl-match-hostname  3.5.0.1     
BeautifulSoup                 3.2.1       
billiard                      3.5.0.3     
cached-property               1.5.1       
cachetools                    2.0.1       
celery                        4.2.1       
certifi                       2018.4.16   
cffi                          1.11.5      
channels                      1.1.8       
chardet                       3.0.4       
Click                         7.0         
colorama                      0.2.5       
commonmark                    0.8.1       
ConfigArgParse                0.10.0      
configobj                     5.0.6       
configparser                  3.7.4       
constantly                    15.1.0      
construct                     2.5.3       
contextlib2                   0.5.5       
coverage                      4.5.3       
cryptography                  1.2.3       
cycler                        0.10.0      
daphne                        1.4.2       
defer                         1.0.4       
Django                        1.11.14     
django-allauth                0.31.0      
django-appconf                1.0.2       
django-autoslug               1.9.3       
django-braces                 1.12.0      
django-cachalot               2.0.2       
django-cache-machine          0.9.1       
django-cache-manager          0.5         
django-celery                 3.1.17      
django-celery-beat            1.0.1       
django-cities                 0.4.2       
django-cities-light           3.3.0       
django-cleanup                0.4.2       
django-compressor             2.1         
django-cors-headers           2.1.0       
django-debug-toolbar          1.11        
django-elasticsearch          0.5         
django-elasticsearch-dsl      0.4.4       
django-extensions             2.0.7       
django-fancy-cronfield        0.2.0       
django-filter                 1.1.0       
django-hvad                   1.8.0       
django-ipware                 2.1.0       
django-oauth-toolkit          0.12.0      
django-password-validation    0.1.1       
django-phonenumber-field      2.2.0       
django-push-notifications     1.1.0       
django-redis                  4.8.0       
django-redis-cache            1.7.1       
django-rest-auth              0.9.0       
django-rest-swagger           0.3.10      
django-silk                   3.0.2       
django-socketio               0.3.9       
django-uuidfield              0.5.0       
django-websocket-redis        0.4.6       
django-yandex-money           1.1.3       
djangorestframework           3.7.7       
djangorestframework-jwt       1.9.0       
djangorestframework-recaptcha 0.2.0       
dnspython                     1.11.1      
docutils                      0.11        
drf-cached-instances          0.3.4       
drf-extensions                0.3.1       
elasticsearch                 5.5.2       
elasticsearch-dsl             5.4.0       
enum34                        1.1.2       
execnet                       1.6.0       
Faker                         0.9.1       
flower                        0.9.2       
freezegun                     0.3.8       
funcsigs                      1.0.2       
future                        0.16.0      
futures                       3.2.0       
gevent                        1.2.2       
gevent-socketio               0.2.1       
gevent-websocket              0.9.5       
google-api-core               1.1.2       
google-api-python-client      1.5.5       
google-auth                   1.4.1       
google-cloud-speech           0.34.0      
googleapis-common-protos      1.5.3       
gprof2dot                     2017.9.19   
greenlet                      0.4.12      
grpcio                        1.11.0      
html5lib                      0.999       
httplib2                      0.9.2       
hyperlink                     19.0.0      
idna                          2.6         
importlib-metadata            0.18        
incremental                   16.10.1     
infi.clickhouse-orm           1.0.3       
influxdb                      4.1.1       
ipaddress                     1.0.16      
iso8601                       0.1.12      
Jinja2                        2.8         
kiwisolver                    1.0.1       
kombu                         4.2.1       
letsencrypt                   0.4.1       
librabbitmq                   1.6.1       
livereload                    2.6.0       
lockfile                      0.8         
lxml                          4.3.3       
Markdown                      3.0.1       
MarkupSafe                    0.23        
matplotlib                    2.2.2       
meld3                         1.0.2       
mixer                         6.1.3       
mkdocs                        0.14.0      
mock                          1.3.0       
more-itertools                5.0.0       
msgpack-python                0.4.8       
mutagen                       1.22        
MySQL-python                  1.2.5       
ndg-httpsclient               0.4.0       
numpy                         1.13.1      
oauth2client                  4.0.0       
oauthlib                      2.0.1       
packaging                     19.0        
pam                           0.1.4       
pandas                        0.24.2      
parsedatetime                 1.4         
pathlib2                      2.3.3       
pbr                           1.8.0       
pefile                        2016.3.28   
pep8                          1.7.0       
pexpect                       3.1         
phonenumbers                  8.10.5      
pika                          0.11.2      
Pillow                        3.4.2       
pip                           19.1.1      
piston-mini-client            0.7.5       
pluggy                        0.12.0      
posix-ipc                     1.0.4       
progressbar2                  3.11.0      
protobuf                      3.5.2.post1 
psutil                        3.4.2       
py                            1.8.0       
pyasn1                        0.4.2       
pyasn1-modules                0.2.1       
pycparser                     2.18        
pycrypto                      2.6.1       
pydevd                        1.0.0       
Pygments                      2.1.3       
PyHamcrest                    1.9.0       
PyICU                         2.3.1       
PyJWT                         1.4.2       
PyMySQL                       0.9.3       
pyOpenSSL                     0.15.1      
pyparsing                     2.2.0       
pyRFC3339                     1.0         
pyserial                      2.6         
pytest                        4.6.4       
pytest-cov                    2.7.1       
pytest-django                 3.4.8       
pytest-forked                 1.0.2       
pytest-mock                   1.10.4      
pytest-xdist                  1.29.0      
python-crontab                1.9.3       
python-dateutil               2.6.0       
python-logstash               0.4.6       
python-memcached              1.58        
python-mpd                    0.3.0       
python-openid                 2.2.5       
python-ptrace                 0.9.1       
python-social-auth            0.2.21      
python-utils                  2.0.1       
python2-pythondialog          3.3.0       
pytz                          2018.3      
pyxdg                         0.25        
PyYAML                        3.12        
rcssmin                       1.0.6       
readthedocs-build             2.0.5       
readthedocs-sphinx-ext        0.5.16      
recommonmark                  0.5.0       
redis                         2.10.6      
reportlab                     3.0         
requests                      2.18.4      
requests-oauthlib             0.7.0       
rest-social-auth              1.3.1       
rjsmin                        1.0.12      
rsa                           3.4.2       
scandir                       1.10.0      
scipy                         0.18.1      
serpy                         0.1.1       
service-identity              16.0.0      
setuptools                    41.0.1      
singledispatch                3.4.0.3     
six                           1.11.0      
snowballstemmer               1.2.1       
social-auth-app-django        2.1.0       
social-auth-core              1.7.0       
Sphinx                        1.3.5       
sphinx-me                     0.3         
sphinx-rtd-theme              0.4.2       
sqlparse                      0.2.2       
ssh-import-id                 3.21        
subprocess32                  3.5.2       
supervisor                    3.3.1       
svgpath2mpl                   0.2.1       
text-unidecode                1.2         
tornado                       4.5.2       
tqdm                          4.19.4      
Twisted                       18.9.0      
txaio                         2.5.2       
typing                        3.6.4       
ujson                         1.35        
Unidecode                     0.4.19      
uritemplate                   3.0.0       
urllib3                       1.22        
uuid                          1.30        
uWSGI                         2.0.18      
vine                          1.1.4       
virtualenv                    15.0.1      
virtualenv-clone              0.5.3       
wcwidth                       0.1.7       
websocket-client              0.44.0      
wheel                         0.32.3      
wsaccel                       0.6.2       
yandex-checkout               1.0.0       
zipp                          0.5.1       
zope.component                4.2.2       
zope.event                    4.2.0       
zope.hookable                 4.0.4       
zope.interface                4.6.0       

```

**Comment by nicoddemus** [comment]
Hi @clockdev,

Hmmm it is kind of hard to follow by the example code you posted. Can you please reduce the code to a small reproducible example that we can run? Also, what do you mean by "inherited"? I ask because I don't see any classes in your code at all. Fixtures are explicitly requested by test functions by declaring them in the parameter list, no inheritance happens implicitly.

**Comment by clockdev** [comment]
Hi @nicoddemus,
I want to implement a check for permissions to execute the API request. I use the fixture as a mock injection where permissions are written to the database. I expect the fixture to be used in the database at the test function level. Where we call fixture, there should be a record of rights in the database. Where not calling in the database should be empty.

By inheritance, I mean that the behaviour of the fixtures is not subject to the above logic. Maybe in my case just need mock instead of fixtures?

But I am interested in one question:
Why when performing tests without fixtures, tests with fixtures do not use the fixture itself? And vice versa - tests with fixtures use fixtures when they start, but further tests without fixtures also use it. In the second case, I can understand that the database may not be cleared. But the first case is still not clear to me.

In the first case, the fixtures are not applied at all. In the second case, they are simply not cleared after the end of each function.

**Comment by Zac-HD** [comment]
I think mocks probably _are_ a better choice for this use-case.
